pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -54,29 +55,61 @@ type AmbientIndex struct {
 	byService map[string][]*model.WorkloadInfo
 	// byPod indexes by Pod IP address.
 	byPod map[string]*model.WorkloadInfo
+	// services are indexed by the network/clusterIP
+	// TODO: GregHanson index by service hostname+port?
+	services map[string]*workloadapi.Service
```

Comment:
nit; might be cleaner to create an internal type `NetworkAddress` to be the key here

Reply:
I'm wondering if `ResourceName()` should return a `NetworkAddress` struct, and that struct can have a `ToString()` method for when it's actually needed as a string (i.e., when sent as xds name across the wire, but otherwise internally always the correct type)

Reply:
byPod is the one that should eventually support the UID look up right?  So _eventually_ are two map lookups performed, once for UID and if none found, again for network/IP?  In which case, does it make sense to have one map[string]Workload, that can handle both lookups?  If so, I agree, it makes sense that ResourceName() return a struct containing those 3 fields (network, ip, uid)

Reply:
per the design doc, we can only look up workloads on-demand by `network/IP`. the internal storage for the UID ones could be string-keyed imo, and the network/IP ones keyed by the struct.

when we look up by network/IP, when we sent across the wire over XDS it will use the UID name, although not in this first version of the PR (to keep scope down)